src/Generator/TypeGenerator.php

```diff
@@ -49,88 +51,160 @@ final class TypeGenerator implements GeneratorInterface
+        ReflectionNamedType $type,
+        ?ReflectionClass $currentClass
+    ): string {
+        $lowerCaseName = strtolower($type->getName());
 
-        if ($wasTrimmed && $isInternalPhpType) {
-            throw new InvalidArgumentException(sprintf(
-                'Provided type "%s" is an internal PHP type, but was provided with a namespace separator prefix',
-                $type
-            ));
+        if ('mixed' === $lowerCaseName || 'null' === $lowerCaseName) {
+            // `mixed` and `null` are implicitly nullable, therefore we need to skip adding nullability markers to it
+            return $lowerCaseName;
         }
 
-        if ($nullable && $isInternalPhpType && 'void' === strtolower($trimmedType)) {
-            throw new InvalidArgumentException(sprintf('Provided type "%s" cannot be nullable', $type));
+        $nullabilityMarker = $type->allowsNull()
+            ? '?'
+            : '';
+
+        if ('self' === $lowerCaseName && $currentClass) {
+            return $nullabilityMarker . $currentClass->getName();
         }
 
-        $instance = new self();
+        if ('parent' === $lowerCaseName && $currentClass && $parentClass = $currentClass->getParentClass()) {
+            return $nullabilityMarker . $parentClass->getName();
+        }
 
-        $instance->type              = $trimmedType;
-        $instance->nullable          = $nullable;
-        $instance->isInternalPhpType = $isInternalPhpType;
+        return $nullabilityMarker . $type->getName();
+    }
 
-        return $instance;
+    /**
+     * @throws InvalidArgumentException
+     *
+     * @psalm-pure
+     */
+    public static function fromTypeString(string $type): self
+    {
+        [$nullable, $trimmedNullable] = self::trimNullable($type);
+        $types = array_map([AtomicType::class, 'fromString'], explode('|', $trimmedNullable));
+
+        usort(
+            $types,
+            static fn(AtomicType $left, AtomicType $right): int => [$left->sortIndex,
+                                                                    $left->type] <=> [$right->sortIndex, $right->type]
+        );
+
+        if (1 === count($types)) {
+            $types[0]->assertCanBeAStandaloneType();
+
+            if ($nullable) {
+                $types[0]->assertCanBeStandaloneNullable();
+            }
+        } else {
+            if ($nullable) {
+                throw new InvalidArgumentException(sprintf(
+                    'Type "%s" is a union type, and therefore cannot be also marked nullable with the "?" prefix',
+                    $type
+                ));
+            }
+
+            foreach ($types as $index => $atomicType) {
+                $atomicType->assertCanUnionWith(array_diff_key($types, array_flip([$index])));
+            }
+        }
+
+        return new self($types, $nullable);
     }
 
-    private function __construct()
+    /**
+     * @var AtomicType[]                     $types
+     *
```

Comment:
Is it standard to have an empty line between standard and Psalm-specific annotations? (mainly curious, so I can be consistent!)

Reply:
I got used to this from `doctrine/coding-standard`, where I group annotations by tool/rationale.

For example, `@param` annotations are split from `@return` by at least one empty line.

No convention though: if needed, can propose it for our own CS, but I'm not attached to this.